### ISSUE TITLE
Cabal complains about `TypeAbstractions`

### ISSUE THREAD

**Issue by LeventErkok** [haskell/cabal#9496]
As I understand it GHC 9.8.1 should work well with cabal 3.10.2.1, and indeed I've those versions:

```
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 9.8.1
$ cabal --version
cabal-install version 3.10.2.1
compiled using version 3.10.2.0 of the Cabal library
```

But I'm getting:

```
$ cabal check
Warning: The following warnings are likely to affect your build negatively:
Warning: Unknown extensions: TypeAbstractions, TypeAbstractions,
TypeAbstractions, TypeAbstractions, TypeAbstractions, TypeAbstractions
...
```

The cabal file under question is: https://github.com/LeventErkok/sbv/blob/master/sbv.cabal

`TypeAbstractions` is in a conditional stanza (https://github.com/LeventErkok/sbv/blob/48f39ce0666b976ebba559cc27a55a3c41d2fcc0/sbv.cabal#L79-L81), but I don't think that matters. Cabal still complains the same if I move it out of the conditional part.

**Comment by Mikolaj** [comment]
Let me mark it as 3.10 backport so that we don't forget to decide.

**Comment by ffaf1** [comment]
Depending on how important this is re: blessed-by-ghcup, this *can* be backported to 3.10, without adding type constructors, just modifying the logic of `check`.

**Comment by Mikolaj** [comment]
Andrea, you are right we can't just add a new type constructor. This is a shame and one more reason to review why we missed the extension. IMHO, this is worth a hack that Francesco suggests. GHC warns to add that extension, so more and more people are going to have it, so the warning at each `cabal build`, repeated for each component, is going to be annoying.

**Comment by andreabedini** [comment]
> Depending on how important this is re: blessed-by-ghcup, this _can_ be backported to 3.10, without adding type constructors, just modifying the logic of `check`.

Right. W.r.t to ghcup, ghcup only ships the executable binary so there's no issues of compatibility there. But we need to be careful not to add a breaking change to the 3.10 branch, otherwise we risk forgetting about it and mixing it up with other backports.

@Mikolaj yeah, unfortunately it is. From https://pvp.haskell.org/

```
Breaking change. If any entity was removed, or the types of any entities or the definitions of datatypes or classes were changed, or orphan instances were added or any instances were removed, then the new A.B MUST be greater than the previous A.B. Note that modifying imports or depending on a newer version of another package may cause extra orphan instances to be exported and thus force a major version change.
```

![image](https://github.com/haskell/cabal/assets/69135/357f5a5b-932c-414f-baa5-1f615fa1e1e0)

The reason is simple, if a dependent package case splits on the data type, it will either fail to build or worse get an exception at runtime.

I don't have any issue with modifying how cabal check works.

**Comment by Mikolaj** [comment]
Right. I think @ffaf1 might have meant that if the "recommended" status bestowed by ghcup, which we so eagerly seek, depends on handling the extension, that's one more argument to make the extra effort and hack this in 3.10.

**Comment by Mikolaj** [comment]
Re 3.10 postmortem: and how about the GHC test https://gitlab.haskell.org/ghc/ghc/-/blob/master/testsuite/tests/driver/T4437.hs? How come it did not detect the missing extension? Is it not run by GHC CI? @bgamari, @mpickering?

**Comment by Mikolaj** [comment]
Wow, `TypeAbstractions` is marked there as `expectedGhcOnlyExtensions`! Is cabal not supposed to support it? What a mess!

**Comment by philderbeast** [comment]
Something I was looking into for #9435 might help:

https://github.com/haskell/cabal/blob/2dad49aae0d691896df3ad3c79c57ae55637a54e/Cabal-tests/tests/UnitTests/Distribution/Simple/Program/GHC.hs#L80-L89

Here's a sequence of commands that shows `-XTypeAbstractions` as added:

```
$ ghc-9.6.3 --show-options | sort > 9.6.3.txt
$ ghc-9.8.1 --show-options | sort > 9.8.1.txt
$ diff -u 9.6.3.txt 9.8.1.txt | grep -E "^\+-X" | sed -E 's/\+(.*)$/, "\1"/'
, "-XExtendedLiterals"
, "-XNoExtendedLiterals"
, "-XNoTypeAbstractions"
, "-XTypeAbstractions"
```

Are option changes only ever shipped the first of a GHC series, in the `ghc-x.y.1` version?

**Comment by Mikolaj** [comment]
@int-index: thank you very much for the clarification. It seems we missed to double-check from the cabal side, as well. I've added a hint to use `ghc --supported-languages` to double check cabal code vs GHC output to out release checklist, so it should be more obvious what to do for the 3.12.1 release. 

Also, it seems the communication with GHC devs that grace our cabal dev meetings failed in this respect --- I'm being told we got distracted by long term automation and decoupling plans discussion and never got back to, e.g., cross-checking #9203 and a similar integration ticket on the GHC side, if there was any this time (there was for previous GHC releases).

@bgamari says "I think all we can do is to potentially make the T4437 fail during release builds" --- that sounds great. Let's not close this here cabal ticket before it's resolved.

All in all, the postmortem up to this point indicates it's not a major process or communication break-down, but just an unfortunate combination of little problems. Thank you all for keeping at it and let's also work on the long term automation/decoupling solutions!
